File: crates/sim-terrain/src/roads.rs
//! 道路の敷設。
//!
//! 仕様 `docs/spec/03-terrain.md` 2.6 節・6 節（段階 12）。
//!
//! フルの地形グリッド（5 km マップで最大 2500² セル）上で直接 Dijkstra を
//! 回すと重いので、粗いグリッド（既定 8 m/セル相当）に落として経路を求め、
//! 結果をラスタライズしてから元のグリッドへ焼き戻す。拠点間は勾配コストに
//! 従うので、平地を通り、川は流量の小さい場所（狭い場所）を選んで渡る
//! 傾向が自然に出る。交点は自動的に橋になる。

use crate::{Surface, Terrain, TerrainParams, SURFACE_EFFECTS};
use sim_math::{Purpose, Rng};
use std::cmp::Reverse;
use std::collections::BinaryHeap;

/// 粗いグリッド 1 セルあたりの元グリッドのセル数。
const COARSEN: i32 = 4;
/// 道路の幅（片側、fine セル単位）。1 なら中心 ±1 セル = 3 セル幅。
const ROAD_HALF_WIDTH: i32 = 1;

/// 道路を敷設し、`t.surface` に `Road` / `Bridge` を焼き込む。
pub fn lay_roads(t: &mut Terrain, p: &TerrainParams) {
    if p.road_count == 0 {
        return;
    }
    let dim = t.dim as i32;
    if dim < COARSEN * 4 {
        return; // マップが小さすぎて粗視化する意味がない
    }
    let coarse_dim = (dim / COARSEN).max(2);

    let coarse_cost = build_coarse_cost(t, coarse_dim);
    let endpoints = pick_endpoints(t, p, coarse_dim);

    for (a, b) in endpoints {
        if let Some(path) = shortest_path(&coarse_cost, coarse_dim, a, b) {
            rasterize_path(t, &path, coarse_dim);
        }
    }
}

/// 粗いグリッド 1 セルの通行コスト。値が大きいほど通りたくない。
fn build_coarse_cost(t: &Terrain, coarse_dim: i32) -> Vec<u32> {
    let mut cost = vec![0u32; (coarse_dim * coarse_dim) as usize];
    for cy in 0..coarse_dim {
        for cx in 0..coarse_dim {
            let (fx, fy) = coarse_to_fine_center(cx, cy, t.dim as i32);
            let idx = t.idx(fx as u32, fy as u32);
            let surface = Surface::from_u8(t.surface[idx]);
            let eff = &SURFACE_EFFECTS[surface as usize];

            // move_mult が高い（歩きやすい）ほどコストを下げる
            let mut c: i32 = 1200 - eff.move_mult as i32;
            c = c.max(50);

            let depth_cm = t.water[idx] as i32 * 10;
            if depth_cm > 0 {
                // 水域は渡橋コストとして重く見積もる。深いほど橋が長く高くつく
                c += 500 + depth_cm * 3;
            }

            // 粗いセル内の標高差を大まかな勾配コストにする
            let h_here = t.height_at_cell(fx, fy) as i32;
            let h_east = t.height_at_cell(fx + COARSEN, fy) as i32;
            let h_south = t.height_at_cell(fx, fy + COARSEN) as i32;
            let slope = (h_here - h_east).abs().max((h_here - h_south).abs());
            c += slope / 4;

            cost[(cy * coarse_dim + cx) as usize] = c.max(1) as u32;
        }
    }
    cost
}

#[inline]
fn coarse_to_fine_center(cx: i32, cy: i32, fine_dim: i32) -> (i32, i32) {
    let fx = (cx * COARSEN + COARSEN / 2).min(fine_dim - 1);
    let fy = (cy * COARSEN + COARSEN / 2).min(fine_dim - 1);
    (fx, fy)
}

/// 道路の始点・終点ペアを決める。地図の対辺同士を、辺に沿ってずらしながら結ぶ。
fn pick_endpoints(t: &Terrain, p: &TerrainParams, coarse_dim: i32) -> Vec<(i32, i32)> {
    let mut rng = Rng::stream(p.seed, 0, Purpose::Terrain, 0xC0AD);
    let mut pairs = Vec::new();
    let last = coarse_dim - 1;

    for i in 0..p.road_count {
        let jitter_a = rng.range(0, coarse_dim);
        let jitter_b = rng.range(0, coarse_dim);
        let (a, b) = if i % 2 == 0 {
            // 西 → 東
            (jitter_a * coarse_dim, last * coarse_dim + jitter_b)
        } else {
            // 北 → 南
            (jitter_a, last * coarse_dim + jitter_b)
        };
        pairs.push((a, b));
    }

    let _ = t;
    pairs
}

/// 粗いグリッド上の Dijkstra 探索。決定論を保つため、優先度キューのタイブレークに
/// ノード index を使う（コストが同じでも走査順に依らず同じ経路が選ばれる）。
fn shortest_path(cost: &[u32], coarse_dim: i32, start: i32, goal: i32) -> Option<Vec<i32>> {
    let n = (coarse_dim * coarse_dim) as usize;
    if start < 0 || goal < 0 || start as usize >= n || goal as usize >= n {
        return None;
    }

    let mut dist = vec![u32::MAX; n];
    let mut prev = vec![-1i32; n];
    let mut heap: BinaryHeap<Reverse<(u32, i32)>> = BinaryHeap::new();

    dist[start as usize] = 0;
    heap.push(Reverse((0, start)));

    while let Some(Reverse((d, idx))) = heap.pop() {
        if idx == goal {
            break;
        }
        if d > dist[idx as usize] {
            continue;
        }
        let cx = idx % coarse_dim;
        let cy = idx / coarse_dim;
        for (dx, dy) in [(1, 0), (-1, 0), (0, 1), (0, -1)] {
            let nx = cx + dx;
            let ny = cy + dy;
            if nx < 0 || ny < 0 || nx >= coarse_dim || ny >= coarse_dim {
                continue;
            }
            let nidx = ny * coarse_dim + nx;
            let nd = d + cost[nidx as usize];
            if nd < dist[nidx as usize] {
                dist[nidx as usize] = nd;
                prev[nidx as usize] = idx;
                heap.push(Reverse((nd, nidx)));
            }
        }
    }

    if dist[goal as usize] == u32::MAX {
        return None;
    }

    let mut path = vec![goal];
    let mut cur = goal;
    while cur != start {
        cur = prev[cur as usize];
        if cur < 0 {
            return None;
        }
        path.push(cur);
    }
    path.reverse();
    Some(path)
}

/// 粗い経路を fine グリッドへラスタライズし、`Road` / `Bridge` を焼く。
fn rasterize_path(t: &mut Terrain, coarse_path: &[i32], coarse_dim: i32) {
    let fine_dim = t.dim as i32;
    let mut fine_points = Vec::with_capacity(coarse_path.len());
    for &c in coarse_path {
        let cx = c % coarse_dim;
        let cy = c / coarse_dim;
        fine_points.push(coarse_to_fine_center(cx, cy, fine_dim));
    }

    for w in fine_points.windows(2) {
        let (x0, y0) = w[0];
        let (x1, y1) = w[1];
        for (lx, ly) in bresenham(x0, y0, x1, y1) {
            stamp_road(t, lx, ly, fine_dim);
        }
    }
    if let Some(&(x, y)) = fine_points.first() {
        stamp_road(t, x, y, fine_dim);
    }
}

/// 1 点の周囲 `ROAD_HALF_WIDTH` セルを道路（水域なら橋）にする。
fn stamp_road(t: &mut Terrain, cx: i32, cy: i32, fine_dim: i32) {
    for dy in -ROAD_HALF_WIDTH..=ROAD_HALF_WIDTH {
        for dx in -ROAD_HALF_WIDTH..=ROAD_HALF_WIDTH {
            let x = cx + dx;
            let y = cy + dy;
            if x < 0 || y < 0 || x >= fine_dim || y >= fine_dim {
                continue;
            }
            let idx = t.idx(x as u32, y as u32);
            t.surface[idx] = if t.water[idx] > 0 {
                Surface::Bridge as u8
            } else {
                Surface::Road as u8
            };
        }
    }
}

/// 整数 Bresenham 直線。始点・終点を含む。
fn bresenham(x0: i32, y0: i32, x1: i32, y1: i32) -> Vec<(i32, i32)> {
    let mut points = Vec::new();
    let dx = (x1 - x0).abs();
    let dy = -(y1 - y0).abs();
    let sx = if x0 < x1 { 1 } else { -1 };
    let sy = if y0 < y1 { 1 } else { -1 };
    let mut err = dx + dy;
    let (mut x, mut y) = (x0, y0);

    loop {
        points.push((x, y));
        if x == x1 && y == y1 {
            break;
        }
        let e2 = 2 * err;
        if e2 >= dy {
            err += dy;
            x += sx;
        }
        if e2 <= dx {
            err += dx;
            y += sy;
        }
    }
    points
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{generate, TerrainParams};

    fn params(seed: u64, road_count: u16) -> TerrainParams {
        TerrainParams {
            seed,
            size_m: 800,
            cell_m: 2,
            relief: 400,
            thermal_iterations: 5,
            road_count,
            ..Default::default()
        }
    }

    #[test]
    fn roads_are_deterministic() {
        let a = generate(&params(11, 2));
        let b = generate(&params(11, 2));
        assert_eq!(a.hash(), b.hash());
    }

    #[test]
    fn zero_roads_means_no_road_surface() {
        let t = generate(&params(12, 0));
        assert!(!t.surface.contains(&(Surface::Road as u8)));
    }

    #[test]
    fn requested_roads_produce_road_or_bridge_tiles() {
        let t = generate(&params(13, 2));
        let road_tiles = t
            .surface
            .iter()
            .filter(|&&s| s == Surface::Road as u8 || s == Surface::Bridge as u8)
            .count();
        assert!(road_tiles > 0, "道路が 1 タイルも生成されていない");
    }

    #[test]
    fn bresenham_includes_both_endpoints() {
        let pts = bresenham(0, 0, 5, 3);
        assert_eq!(pts[0], (0, 0));
        assert_eq!(*pts.last().unwrap(), (5, 3));
    }

    #[test]
    fn shortest_path_finds_a_route_on_uniform_cost() {
        let dim = 5;
        let cost = vec![10u32; (dim * dim) as usize];
        let path = shortest_path(&cost, dim, 0, dim * dim - 1).expect("経路が見つからない");
        assert_eq!(path.first().copied(), Some(0));
        assert_eq!(path.last().copied(), Some(dim * dim - 1));
    }

    #[test]
    fn shortest_path_avoids_expensive_cells() {
        // 中央の列 (x=3) に、1 か所だけ隙間 (y=0) を残した壁を作る 7x7 グリッド。
        // 中段 (y=3) から出発して反対側へ抜けるには、壁のない y=0 まで
        // 迂回するはずで、壁のセルは一切通らない。
        let dim = 7;
        let mut cost = vec![10u32; (dim * dim) as usize];
        for y in 1..dim {
            cost[(y * dim + 3) as usize] = 1_000_000;
        }
        let start = 3 * dim; // (x=0, y=3)
        let goal = 3 * dim + dim - 1; // (x=6, y=3)
        let path = shortest_path(&cost, dim, start, goal).unwrap();
        let crosses_wall = path.iter().any(|&idx| idx % dim == 3 && idx / dim != 0);
        assert!(!crosses_wall, "壁を突っ切っている: {path:?}");
        // 隙間 (x=3, y=0) は通っているはず
        assert!(path.contains(&3));
    }
}

File: web/src/terrain/roads.ts
/**
 * 道路の敷設。仕様 `docs/spec/03-terrain.md` 2 節（段階 12）。
 *
 * フルの地形グリッド（5 km マップで最大 2500² セル）上で直接 Dijkstra を
 * 回すと重いので、粗いグリッド（既定 8 m/セル相当）に落として経路を求め、
 * 結果をラスタライズしてから元のグリッドへ焼き戻す。拠点間は勾配コストに
 * 従うので、平地を通り、川は狭い場所を選んで渡る傾向が自然に出る。
 * 水域との交点は自動的に橋になる。
 *
 * 道路は `overlay` グリッドに書く。地質と植生には触らない——路面の下の
 * 土は土のままで、工兵が道を掘り返せば元の地質が戻る。
 */

import { basePassability } from "./effects";
import { Overlay } from "./types";
import type { GenerationResult } from "./generate";
import { Domain, Stream, derive32 } from "./rng";

/** 粗いグリッド 1 セルあたりの元グリッドのセル数。 */
const COARSEN = 4;
/** 道路の幅（片側、fine セル単位）。1 なら中心 ±1 セル = 3 セル幅。 */
const ROAD_HALF_WIDTH = 1;

/** 道路を敷設し、`overlay` に `ROAD` / `BRIDGE` を焼き込む。 */
export function layRoads(t: GenerationResult, roadCount: number): void {
  if (roadCount <= 0) return;
  const dim = t.dim;
  if (dim < COARSEN * 4) return; // マップが小さすぎて粗視化する意味がない
  const coarseDim = Math.max(2, Math.floor(dim / COARSEN));

  const cost = buildCoarseCost(t, coarseDim);
  const rng = new Stream(derive32(t.seed, Domain.ROADS));
  const last = coarseDim - 1;

  for (let i = 0; i < roadCount; i++) {
    const jitterA = rng.range(0, coarseDim);
    const jitterB = rng.range(0, coarseDim);
    // 偶数本目は西→東、奇数本目は北→南。地図の対辺同士を結ぶ
    const a = i % 2 === 0 ? jitterA * coarseDim : jitterA;
    const b = last * coarseDim + jitterB;
    const path = shortestPath(cost, coarseDim, a, b);
    if (path) rasterizePath(t, path, coarseDim);
  }
}

/** 粗いグリッド 1 セルの通行コスト。値が大きいほど通りたくない。 */
function buildCoarseCost(t: GenerationResult, coarseDim: number): Uint32Array {
  const cost = new Uint32Array(coarseDim * coarseDim);
  const fineDim = t.dim;
  for (let cy = 0; cy < coarseDim; cy++) {
    for (let cx = 0; cx < coarseDim; cx++) {
      const [fx, fy] = coarseToFineCenter(cx, cy, fineDim);
      const i = fy * fineDim + fx;
      const depth = t.water[i]!;

      // 歩きやすいほどコストを下げる。通行不能セルは事実上の壁
      const pass = basePassability(t.ground[i]!, t.vegetation[i]!, depth, Overlay.NONE);
      let c = pass === 0 ? 100_000 : Math.max(50, 1200 - Math.floor((pass * 1150) / 255));

      if (depth > 0) {
        // 水域は渡橋コストとして重く見積もる。深いほど橋が長く高くつく
        c += 500 + depth * 3;
      }

      // 粗いセル内の標高差を大まかな勾配コストにする
      const hHere = t.height[i]!;
      const hEast = t.height[fy * fineDim + Math.min(fineDim - 1, fx + COARSEN)]!;
      const hSouth = t.height[Math.min(fineDim - 1, fy + COARSEN) * fineDim + fx]!;
      c += Math.floor(Math.max(Math.abs(hHere - hEast), Math.abs(hHere - hSouth)) / 4);

      cost[cy * coarseDim + cx] = Math.max(1, c);
    }
  }
  return cost;
}

function coarseToFineCenter(cx: number, cy: number, fineDim: number): [number, number] {
  return [
    Math.min(fineDim - 1, cx * COARSEN + (COARSEN >> 1)),
    Math.min(fineDim - 1, cy * COARSEN + (COARSEN >> 1)),
  ];
}

/**
 * 粗いグリッド上の Dijkstra 探索。
 *
 * 優先度キューのタイブレークにノード index を使うので、コストが同じでも
 * 走査順に依らず同じ経路が選ばれる。
 */
export function shortestPath(
  cost: Uint32Array,
  coarseDim: number,
  start: number,
  goal: number,
): number[] | null {
  const n = coarseDim * coarseDim;
  if (start < 0 || goal < 0 || start >= n || goal >= n) return null;

  const dist = new Uint32Array(n).fill(0xffffffff);
  const prev = new Int32Array(n).fill(-1);
  const heap = new MinHeap();

  dist[start] = 0;
  heap.push(0, start);

  while (heap.size > 0) {
    const [d, idx] = heap.pop()!;
    if (idx === goal) break;
    if (d > dist[idx]!) continue;
    const cx = idx % coarseDim;
    const cy = (idx / coarseDim) | 0;
    for (let k = 0; k < 4; k++) {
      const nx = cx + (k === 0 ? 1 : k === 1 ? -1 : 0);
      const ny = cy + (k === 2 ? 1 : k === 3 ? -1 : 0);
      if (nx < 0 || ny < 0 || nx >= coarseDim || ny >= coarseDim) continue;
      const nidx = ny * coarseDim + nx;
      const nd = d + cost[nidx]!;
      if (nd < dist[nidx]!) {
        dist[nidx] = nd;
        prev[nidx] = idx;
        heap.push(nd, nidx);
      }
    }
  }

  if (dist[goal] === 0xffffffff) return null;
  const path = [goal];
  let cur = goal;
  while (cur !== start) {
    cur = prev[cur]!;
    if (cur < 0) return null;
    path.push(cur);
  }
  path.reverse();
  return path;
}

/** コストとノード index の 2 段キーで並ぶ最小ヒープ。 */
class MinHeap {
  private readonly keys: number[] = [];
  private readonly vals: number[] = [];

  get size(): number {
    return this.keys.length;
  }

  private less(i: number, j: number): boolean {
    return this.keys[i]! < this.keys[j]! || (this.keys[i] === this.keys[j] && this.vals[i]! < this.vals[j]!);
  }

  private swap(i: number, j: number): void {
    [this.keys[i], this.keys[j]] = [this.keys[j]!, this.keys[i]!];
    [this.vals[i], this.vals[j]] = [this.vals[j]!, this.vals[i]!];
  }

  push(key: number, val: number): void {
    this.keys.push(key);
    this.vals.push(val);
    let i = this.keys.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (!this.less(i, parent)) break;
      this.swap(i, parent);
      i = parent;
    }
  }

  pop(): [number, number] | null {
    if (this.keys.length === 0) return null;
    const top: [number, number] = [this.keys[0]!, this.vals[0]!];
    const lastKey = this.keys.pop()!;
    const lastVal = this.vals.pop()!;
    if (this.keys.length > 0) {
      this.keys[0] = lastKey;
      this.vals[0] = lastVal;
      let i = 0;
      for (;;) {
        const l = i * 2 + 1;
        const r = l + 1;
        let best = i;
        if (l < this.keys.length && this.less(l, best)) best = l;
        if (r < this.keys.length && this.less(r, best)) best = r;
        if (best === i) break;
        this.swap(i, best);
        i = best;
      }
    }
    return top;
  }
}

/** 粗い経路を fine グリッドへラスタライズし、道路・橋を焼く。 */
function rasterizePath(t: GenerationResult, coarsePath: number[], coarseDim: number): void {
  const fineDim = t.dim;
  const points = coarsePath.map((c) =>
    coarseToFineCenter(c % coarseDim, (c / coarseDim) | 0, fineDim),
  );
  for (let i = 0; i + 1 < points.length; i++) {
    for (const [lx, ly] of bresenham(points[i]![0], points[i]![1], points[i + 1]![0], points[i + 1]![1])) {
      stampRoad(t, lx, ly);
    }
  }
  if (points.length > 0) stampRoad(t, points[0]![0], points[0]![1]);
}

/** 1 点の周囲 `ROAD_HALF_WIDTH` セルを道路（水域なら橋）にする。 */
function stampRoad(t: GenerationResult, cx: number, cy: number): void {
  for (let dy = -ROAD_HALF_WIDTH; dy <= ROAD_HALF_WIDTH; dy++) {
    for (let dx = -ROAD_HALF_WIDTH; dx <= ROAD_HALF_WIDTH; dx++) {
      const x = cx + dx;
      const y = cy + dy;
      if (x < 0 || y < 0 || x >= t.dim || y >= t.dim) continue;
      const i = y * t.dim + x;
      t.overlay[i] = t.water[i]! > 0 ? Overlay.BRIDGE : Overlay.ROAD;
    }
  }
}

/** 整数 Bresenham 直線。始点・終点を含む。 */
export function bresenham(x0: number, y0: number, x1: number, y1: number): Array<[number, number]> {
  const points: Array<[number, number]> = [];
  const dx = Math.abs(x1 - x0);
  const dy = -Math.abs(y1 - y0);
  const sx = x0 < x1 ? 1 : -1;
  const sy = y0 < y1 ? 1 : -1;
  let err = dx + dy;
  let x = x0;
  let y = y0;
  for (;;) {
    points.push([x, y]);
    if (x === x1 && y === y1) break;
    const e2 = 2 * err;
    if (e2 >= dy) {
      err += dy;
      x += sx;
    }
    if (e2 <= dx) {
      err += dx;
      y += sy;
    }
  }
  return points;
}
